msprime/simulations.py

```diff
@@ -2254,6 +2254,237 @@ def population_size_trajectory(self, steps):
+                mass_migration_objects.append(demo)
+                mass_migration_times.append(demo.time)
+
+        for jj in range(first_step, len(all_steps)):
+            t_j = all_steps[jj]
+            # apply any mass migration events to _this_ time step
+            # so if we sample at this time, we also account for the instantaneous
+            # mass migration events that occur at the same time
+            for mass_mig_t, mass_mig_e in zip(mass_migration_times,
+                                              mass_migration_objects):
+                if mass_mig_t == t_j:
+                    S = np.eye(num_pops, num_pops)
+                    S[mass_mig_e.source, mass_mig_e.dest] = mass_mig_e.proportion
+                    S[mass_mig_e.source, mass_mig_e.source] = 1 - mass_mig_e.proportion
+                    P = np.matmul(P, S)
+            P_out[jj] = P
+
+            # get continuous migration matrix over next interval
+            if jj + 1 < len(all_steps):
+                _, M = self._pop_size_and_migration_at_t(t_j)
+                dt = all_steps[jj + 1] - all_steps[jj]
+                dM = np.diag([sum(s) for s in M])
+                # changes in probabilities due to mass migration at time t_j
+                # advance to next interval time (dt) taking into account continuous mig
+                P = _matrix_exponential(dt*(M-dM)).dot(P)
+                P_out[jj + 1] = P
+
+        return P_out[keep_steps]
+
+    def lineage_indicators(self, samples=None):
+        """
+        Given the sampling configuration, this function determines when lineages are
+        possibly found within each population over epochs defined by demographic events
+        and sampling times. If no sampling configuration is given, we assume we sample
+        lineages from each population at time zero. The samples are specified by a list
+        of msprime Sample objects, so that possible ancient samples may be accounted for.
+
+        :param list samples: A list of msprime Sample objects, which specify their
+            populations and times.
+        :return: Returns a dictionary of times defining the epoch intervals with arrays
+            of length equal to the number of populations and containing 1s or 0s for
+            whether lineages could be found in that population.
+        """
+        num_pops = self.num_populations
+        # strategy: for each sampling time, calculate where each lineage could end
+        # up at each subsequent epoch, using calculate_lineage_probabilities
+        # then concatenate that data into a minimal dictionary
+
+        sample_sets = {}
+        if samples is None:
+            sample_sets[0] = list(range(num_pops))
+        else:
+            for sample in samples:
+                sample_sets.setdefault(sample.time, [0 for _ in range(num_pops)])
+                sample_sets[sample.time][sample.population] = 1
+
+        steps = [e.start_time for e in self.epochs]
```

Comment:
How about adding an epsilon to each start_time here, where `epsilon = np.finfo(float).eps`?